### PR TITLE
[JumpThreading] Convert `s/zext i1` to `select i1` for further unfolding

### PR DESCRIPTION
**Pull request by FLZ101**
Convert a `s/zext i1` (e.g. `%b = zext i1 %a to i32`) to a `select i1` (e.g. `%b = select i1 %a, i32 1, i32 0`) if doing so helps jump threading.

See `llvm/test/Transforms/JumpThreading/szext.ll` for an example.

### CODE REVIEW

File: llvm/lib/Transforms/Scalar/JumpThreading.cpp [hunk 1]
```diff
@@ -2997,6 +3000,83 @@ bool JumpThreadingPass::tryToUnfoldSelectInCurrBB(BasicBlock *BB) {
...
+  SmallVector<Instruction *> ToConvert;
+  for (auto &I : *BB) {
+    using namespace PatternMatch;
+
+    Value *V;
+    if (!match(&I, m_ZExtOrSExt(m_Value(V))) || !V->getType()->isIntegerTy(1))
+      continue;
+
+    // I is only used by Phi
+    Use *U = I.getSingleUndroppableUse();
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/JumpThreading.cpp:3046]: Use hasOneUse() and use_begin() here. You can't use this API like this.

File: llvm/lib/Transforms/Scalar/JumpThreading.cpp [hunk 2]
```diff
@@ -2997,6 +3000,83 @@ bool JumpThreadingPass::tryToUnfoldSelectInCurrBB(BasicBlock *BB) {
...
+  LLVM_DEBUG(dbgs() << "\nconvert-szext-to-select:\n" << *BB << "\n");
+  for (Instruction *I : ToConvert) {
+    auto Ty = I->getType();
+    Value *V1 = isa<SExtInst>(I) ? ConstantInt::getAllOnesValue(Ty)
+                                 : ConstantInt::get(Ty, 1);
+    Value *V2 = ConstantInt::getNullValue(Ty);
+    SelectInst *SI =
+        SelectInst::Create(I->getOperand(0), V1, V2, I->getName(), I);
+    I->replaceAllUsesWith(SI);
+    I->eraseFromParent();
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/JumpThreading.cpp:3074]: If you use make_early_inc_range, I think you can merge both loops.

File: llvm/test/Transforms/JumpThreading/szext.ll [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  %cmp1.i = icmp sgt i32 %a, %b
+  br i1 %cmp1.i, label %compare1.exit, label %cond.false3.i
+
+cond.false3.i:                                    ; preds = %cond.false.i
+  %cmp4.i = icmp slt i32 %c, %d
+  br i1 %cmp4.i, label %compare1.exit, label %cond.false6.i
+
+cond.false6.i:                                    ; preds = %cond.false3.i
+  %cmp7.i = icmp sgt i32 %c, %d
+  %cond.i = zext i1 %cmp7.i to i32
```
Comment (nikic) [inline llvm/test/Transforms/JumpThreading/szext.ll:68]: Missing test for sext?

This test does not look very minimal.

**Review by nikic — COMMENTED** [review]
@dtcxzyw Can you please test this? I'm not sure whether this does more damage than good...

**Comment by FLZ101** [comment]
Thanks for your comments.

I found a better way to implement this feature and will create a new pull request soon.

**Pull request closed without merging**